File: arima/ruby/arima.rb
class ARIMA
    attr_accessor :ar, :ma, :p, :q, :diff, :residuals, :scale_factor, :forecast, :num_parameters
  
    def initialize(p, q, diff)
      @ar = []
      @ma = []
      @p = p
      @q = q
      @diff = diff
      @residuals = []
      @scale_factor = 0.0
      @forecast = []
      @num_parameters = p + q
    end
  
    def fit(data) 
      # Perform necessary calculations and parameter estimation
  
      # Example placeholder code to demonstrate parameter estimation
      # Assumes p = 1, q = 1, and diff = 0 for simplicity
  
      # Estimate AR coefficient
      @ar << 0.5 # Placeholder value
  
      # Estimate MA coefficient
      @ma << 0.3 # Placeholder value
  
      # Estimate residuals
      @residuals = []
      (diff...data.length).each do |i|
        predicted_value = @ar[0] * data[i - 1] + @ma[0] * @residuals[i - diff]
        @residuals << data[i] - predicted_value
      end
  
      # Calculate scale factor
      min_val = data.min
      max_val = data.max
      @scale_factor = max_val - min_val
    end
  
    def forecast(data, steps)
      forecast = []
  
      # Example placeholder code for forecasting
      # Assumes p = 1, q = 1, and diff = 0 for simplicity
  
      steps.times do
        last_index = data.length + forecast.length - 1
  
        # Forecast the next value based on the AR and MA coefficients
        next_value = @ar[0] * data[last_index] + @ma[0] * @residuals[last_index - diff]
  
        # Add the forecasted value to the result array
        forecast << next_value
      end
  
      forecast
    end
  end
  
  # Usage example
  data = [1.0, 2.0, 3.0, 4.0, 5.0]
  p = 1
  q = 1
  diff = 0
  
  # Create and fit ARIMA model
  model = ARIMA.new(p, q, diff)
  model.fit(data)
  
  # Forecast next value
  forecast_steps = 1
  forecast = model.forecast(data, forecast_steps)
  
  # Print the forecasted value
  puts "Next value: #{forecast[0]}"
  

File: arima/javascript/arima.js
class ARIMA {
    constructor(p, q, diff) {
      this.ar = [];
      this.ma = [];
      this.p = p;
      this.q = q;
      this.diff = diff;
      this.residuals = [];
      this.scaleFactor = 0.0;
      this.forecast = [];
      this.numParameters = p + q;
    }
  
    fit(data) { 
      // Perform necessary calculations and parameter estimation
  
      // Example placeholder code to demonstrate parameter estimation
      // Assumes p = 1, q = 1, and diff = 0 for simplicity
  
      // Estimate AR coefficient
      this.ar.push(0.5); // Placeholder value
  
      // Estimate MA coefficient
      this.ma.push(0.3); // Placeholder value
  
      // Estimate residuals
      this.residuals = [];
      for (let i = this.diff; i < data.length; i++) {
        const predictedValue =
          this.ar[0] * data[i - 1] + this.ma[0] * this.residuals[i - this.diff];
        this.residuals.push(data[i] - predictedValue);
      }
  
      // Calculate scale factor
      let minVal = data[0];
      let maxVal = data[0];
      for (let i = 1; i < data.length; i++) {
        const value = data[i];
        if (value < minVal) {
          minVal = value;
        }
        if (value > maxVal) {
          maxVal = value;
        }
      }
      this.scaleFactor = maxVal - minVal;
    }
  
    forecast(data, steps) {
      const forecast = [];
  
      // Example placeholder code for forecasting
      // Assumes p = 1, q = 1, and diff = 0 for simplicity
  
      for (let i = 0; i < steps; i++) {
        const lastIndex = data.length + i - 1;
  
        // Forecast the next value based on the AR and MA coefficients
        const nextValue =
          this.ar[0] * data[lastIndex] +
          this.ma[0] * this.residuals[lastIndex - this.diff];
  
        // Add the forecasted value to the result list
        forecast.push(nextValue);
      }
  
      return forecast;
    }
  }
  
  // Usage example
  const data = [1.0, 2.0, 3.0, 4.0, 5.0];
  const p = 1;
  const q = 1;
  const diff = 0;
  
  // Create and fit ARIMA model
  const model = new ARIMA(p, q, diff);
  model.fit(data);
  
  // Forecast next value
  const forecastSteps = 1;
  const forecast = model.forecast(data, forecastSteps);
  
  // Print the forecasted value
  console.log("Next value:", forecast[0]);
  